deezone/HydroBytes-BaseStation
# Issue-1 create basic web server

resolves #1 

### Description

in an effort to create a basic API application for the Base Station in the HydroBytes project, this PR creates a prototype API that listens to `localhost:8000` and responds with the details of the request received. In the event of a program cancelation (aka ctrl-c) the application should shutdown gracefully by completing any pending requests.

### How to Test

- send request to `http://localhost:8000` with various paths. The response should be:
![image](https://user-images.githubusercontent.com/2119264/103449110-874e4080-4c71-11eb-8b88-c9ca71f417fe.png)

![image](https://user-images.githubusercontent.com/2119264/103449124-a6e56900-4c71-11eb-8a52-0aad3189651f.png)

- [ ] generate load on the application by using "hey" command line utility (see details below). While some of the requests are still processing, terminate the application. Ensure that any of the requests still processing complete based on the output to the terminal:
```
$ go run .
2021/01/01 21:49:14 main : Started
2021/01/01 21:49:14 main : API listening on localhost:8000

$ hey http://localhost:8000
```

- [ ] random requests made to `http://localhost:8000` that is terminated in the middle of the requests:
```
...
2021/01/01 21:50:17 start 503
2021/01/01 21:50:17 start 552
2021/01/01 21:50:17 start 843
2021/01/01 21:50:17 start 598
2021/01/01 21:50:17 start 205
^C2021/01/01 21:50:17 main : Start shutdown
2021/01/01 21:50:20 end 266
2021/01/01 21:50:20 end 420
2021/01/01 21:50:20 end 133
2021/01/01 21:50:20 end 205
2021/01/01 21:50:20 end 463
...

2021/01/01 21:50:20 end 643
2021/01/01 21:50:20 end 891
2021/01/01 21:50:20 end 2
2021/01/01 21:50:20 main : Completed
```

- [ ] note how requests are completed after termination command is issued
```
Summary:
  Total:	6.0374 secs
  Slowest:	3.0157 secs
  Fastest:	3.0058 secs
  Average:	3.0125 secs
  Requests/sec:	33.1268

  Total data:	1900 bytes
  Size/request:	19 bytes

Response time histogram:
  3.006 [1]	|■■
  3.007 [3]	|■■■■■■
  3.008 [1]	|■■
  3.009 [2]	|■■■■
  3.010 [4]	|■■■■■■■■
  3.011 [7]	|■■■■■■■■■■■■■
  3.012 [16]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  3.013 [21]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  3.014 [7]	|■■■■■■■■■■■■■
  3.015 [19]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  3.016 [19]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■


Latency distribution:
  10% in 3.0097 secs
  25% in 3.0113 secs
  50% in 3.0125 secs
  75% in 3.0144 secs
  90% in 3.0151 secs
  95% in 3.0157 secs
  99% in 3.0157 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0013 secs, 3.0058 secs, 3.0157 secs
  DNS-lookup:	0.0007 secs, 0.0000 secs, 0.0031 secs
  req write:	0.0001 secs, 0.0000 secs, 0.0025 secs
  resp wait:	3.0102 secs, 3.0026 secs, 3.0147 secs
  resp read:	0.0009 secs, 0.0000 secs, 0.0052 secs

Status code distribution:
  [200]	100 responses

Error distribution:
  [100]	Get "http://localhost:8000": dial tcp 127.0.0.1:8000: connect: connection refused
```

**References**:
- "hey" (https://github.com/rakyll/hey)
> hey is a tiny program that sends some load to a web application.
```
brew install hey
```
